performance: Speculative block assembly into master

algorandskiy opened the pull request.

yossigi committed:
plumbing for spec block assembly

yossigi committed:
with spec ledger code

yossigi committed:
block as LFE code

yossigi committed:
build verified block + spec ledger

yossigi committed:
checkpoint: making startevaluator compile

yossigi committed:
add new LFE LedgerForEvaluator functionality

yossigi committed:
forward verified txns cache

yossigi committed:
checkpoint: use spec pool

yossigi committed:
fixing tests

yossigi committed:
speculative ledger unittest updates

yossigi committed:
bug fix: subtract round only as fallback

yossigi committed:
add funcs to read create/read spec blocks

yossigi committed:
checkpoint: stop assembly at first new block

yossigi committed:
cancel spec asm on any new block (speculative or not)

yossigi committed:
plumbing to use spec block

yossigi committed:
get branch from ledger

yossigi committed:
release lock when copy fails

yossigi committed:
don't flush changes from shallow pool copy to real pool

yossigi committed:
Merge branch 'master' into yossi/specblockasm

yossigi committed:
add BlockHdrCache method to interface

yossigi committed:
simpler error handling

yossigi committed:
checkpoint: plumbing through agreement

yossigi committed:
checkpoint: MakeValidatedBlockAsLFE --> MakeBlockAsLFE

yossigi committed:
return statedelta

yossigi committed:
checkpoint: adding readlowest to agreement

yossigi committed:
treat 0 timeout as no timeout

yossigi committed:
x

yossigi committed:
zero spec timeout

yossigi committed:
txpool interface

yossigi committed:
Merge branch 'master' into yossi/specblockasm

yossigi committed:
demux event test

yossigi committed:
fix todos

yossigi committed:
Merge branch 'master' into yossi/specblockasm

yossigi committed:
check casting

yossigi committed:
bug fix + unit test

yossigi committed:
speculative txn not in block

yossigi committed:
asm round to next round

yossigi committed:
better member name

yossigi committed:
check good txns

yossigi committed:
better test

yossigi committed:
reset clears speculative results

cce committed:
change the way pendingTxids and pendingTxGroups are copied

yossigi committed:
merge with current master

yossigi committed:
Merge branch 'master' into yossi/specblockasm

yossigi committed:
Merge branch 'master' into yossi/specblockasm

yossigi committed:
Merge branch 'master' into yossi/specblockasm

yossigi committed:
Merge branch 'master' into yossi/specblockasm

nicholasguo committed:
add test

yossigi committed:
bug fix: return correct round number

yossigi committed:
preserve round number response in case of an error

nicholasguo committed:
use round from 2 rounds ago

yossigi committed:
setting clock for first period after boot

yossigi committed:
Merge branch 'yossi/specblockasm' of github.com:algorand/go-algorand into yossi/specblockasm

nicholasguo committed:
optimize

nicholasguo committed:
Merge branch 'yossi/specblockasm' into nguo/spec-asm-race-test

Commit:
Merge pull request #4556 from nicholasguoalgorand/nguo/spec-asm-race-test

spec asm: Add race unit test

nicholasguo committed:
lint

Commit:
Merge remote-tracking branch 'upstream/master' into feature/specblockasm

Commit:
fix linter and msgp

algorandskiy committed:
Merge pull master into feature/specblockasm

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/algorand/go-algorand?pullRequest=4913) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/algorand/go-algorand?pullRequest=4913) before we can accept your contribution.<br/>**2** out of **6** committers have signed the CLA.<br/><br/>:white_check_mark: cce<br/>:white_check_mark: algorandskiy<br/>:x: yossigi<br/>:x: nicholasguo<br/>:x: brianolson<br/>:x: nicholasguoalgorand<br/><hr/>**nicholasguoalgorand** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/algorand/go-algorand?pullRequest=4913) it.</sub>

brianolson committed:
performance: Speculative block assembly (#4861)

* workaround interface changes

* small tweaks, note TODOs

* cleanup
s/OnNewSpeculativeBlock/StartSpeculativeBlockAssembly/g
rm player.SpeculativeAssemblyDeadline

* simplify demux

* test fixes, works at least -short

* comment

* TODO comments

* expand StartSpeculativeBlockAssembly to take the block header digest

* wip: pool rehacking passes tests

* fix TestSpeculativeBlockAssemblyDataRace
better locking in tryReadSpeculativeBlock()
start thread in StartSpeculativeBlockAssembly()

* fix

* simplify SpeculativeBlockAssemblyGraceTime to SpeculativeAsmTimeOffset
remove player.SpeculativeAsmTimeDuration field

* improve speculative block assembly starting
start spec asm at filterTimeout-offset and again at soft-vote
cleanup

* lint cleanup

* nolint comment

* fix copyright header

* Post merge fixes

* fix GetCreatorForRound which was breaking e2e test

* make generate

* review comment

* SpeculativeAsmTimeOffset typo, 400ms

* algod_speculative_assembly_discarded metric

* start speculative assembly on proposal validation

* StartSpeculativeBlockAssembly(...,onlyIfStarted)

* algod_speculative_assembly_discarded_{start,read,ONB}

* code review feedback

* fix a test by ignoring intermittent speculative block action

Co-authored-by: Pavel Zbitskiy <pavel@algorand.com>

brianolson committed:
Merge remote-tracking branch 'origin/master' into specblockasm

brianolson committed:
license fix

cce committed:
Merge pull request #5066 from brianolson/specblockasm-master-2ba3f4b704

Specblockasm master 2ba3f4b704

yossigi committed:
pull from master

yossigi committed:
versioning

yossigi committed:
spec asm params
